#!/bin/bash

#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mail-user=lucille.delisle@epfl.ch
#SBATCH --nodes 1
#SBATCH --ntasks 1
#SBATCH --cpus-per-task 1
#SBATCH --mem 10G
#SBATCH --time 0:30:00
#SBATCH --job-name Plot
#SBATCH --chdir /scratch/ldelisle/ATAC/

# This script to make Figure 3A with pgt
pathForFasta="/home/ldelisle/genomes/fasta/"
genome=mm10

module purge

# Create and activate a conda environment with pygenometracks version 3.3

source ~/miniconda2/etc/profile.d/conda.sh
exists=`conda info --envs | awk '$1=="pgt_3.3"{print}' | wc -l`
if [ $exists -ne 1 ]; then
  conda create -n pgt_3.3 --yes pygenometracks=3.3
fi
conda activate pgt_3.3

# Get the ChIPseq GSM4294458_H3K4me2_ChIP-seq.bedgraph.gz from GEO
wget "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4294458&format=file&file=GSM4294458%5FH3K4me2%5FChIP%2Dseq%2Ebedgraph%2Egz"
# Convert it to bw
gunzip GSM4294458_H3K4me2_ChIP-seq.bedgraph.gz
bedGraphToBigWig GSM4294458_H3K4me2_ChIP-seq.bedgraph ${pathForFasta}${genome}.fa.fai GSM4294458_H3K4me2_ChIP-seq.bw

# Get the gtf used for RNAseq analysis
wget https://zenodo.org/record/3820860/files/mergeGenesOfFilteredTranscriptsOfMus_musculus.GRCm38.93_ExonsOnly_UCSC.gtf.gz?download=1 -O mergeGenesOfFilteredTranscriptsOfMus_musculus.GRCm38.93_ExonsOnly_UCSC.gtf.gz

# Create a small bed with protein_coding genes around EC
if [ ! -e prot_cod_genes_around_EC.bed ]; then
  # Select only exons from the gtf
  zcat mergeGenesOfFilteredTranscriptsOfMus_musculus.GRCm38.93_ExonsOnly_UCSC.gtf.gz | awk -F '\t' '$1~/^##/{print}$1=="chr15"&&$4<105000000&&$5>101000000&&$9~/"protein_coding"/{print}' > prot_cod_genes_around_EC.gtf

  # Convert the gtf to bed12
  python fromgtfTobed12.py --output prot_cod_genes_around_EC.bed --mergeTranscriptsAndOverlappingExons prot_cod_genes_around_EC.gtf
fi

# Hoxc genes are very close one to the other, Hoxc will be shorten to c
if [ ! -e prot_cod_genes_around_EC_short.bed ]; then
  cat prot_cod_genes_around_EC.bed | awk -v OFS="\t" '{gsub("Hoxc", "c", $4); print}' > prot_cod_genes_around_EC_short.bed
fi

# The configuration file for pgt is created:
output="figure3A"
max_value=35
echo "
[x-axis]
where = top
[enhancers]
file = EC.bed
color = bed_rgb
border_color = none
" > ${output}.ini
for tissue in FDE HDE; do
  echo "[bigwig merged]
file = ATAC_E14_${tissue}_neq2/normalized_ATAC_E14_${tissue}_neq2.bw
title = ATAC E14.5 ${tissue}
height = 3
color = grey
min_value = 0
max_value = ${max_value}
number_of_bins = 500
summary_method = max
" >> ${output}.ini
  for rep in 1 2; do
    echo "[narrowPeak${rep}]
file = ATAC_E14_${tissue}_Rep${rep}/ATAC_E14_${tissue}_Rep${rep}_peaks.narrowPeak
type = box
show_labels = false
color = grey
line_width = 0.2
height = 0.3
" >> ${output}.ini
  done
done
echo "[bigwig norm2]
file = bedGraphs/ATAC_E13_FB_macs_likeATAC_norm2.bw
title = ATAC E13.5 FB
height = 3
color = black
min_value = 0
max_value = ${max_value}
number_of_bins = 500
summary_method = max

[narrowPeakFB]
file = ATAC_E13_FB/ATAC_E13_FB_peaks.narrowPeak
type = box
show_labels = false
color = black
line_width = 0.2
height = 0.3

[bigwig ChIP]
file = GSM4294458_H3K4me2_ChIP-seq.bw
title = H3K4me2 E10.5 FLE
height = 3
color = blue
min_value = 0
number_of_bins = 500
summary_method = max

[genes]
file = prot_cod_genes_around_EC_short.bed
style = UCSC
height = 1
color = black
display = interleaved

[scalebar]
file_type = scalebar
" >> ${output}.ini
mkdir -p pgt_outputs

# The svg is created
pgt --tracks ${output}.ini --region chr15:102700000-103050000 -o pgt_outputs/${output}.svg --trackLabelFraction 0.3 --fontSize 15

# The output has been modified by inkscape to put the labels at the desired place and put vertical area under the EC regions.

# For figure 3B the first track was generated by pgt:
output=forConservation
max_value=35
tissue=FDE
echo "[enhancers]
file = EC.bed
border_color = bed_rgb
color = none
line_width = 3
[bigwig merged]
file = ATAC_E14_${tissue}_neq2/normalized_ATAC_E14_${tissue}_neq2.bw
title = ATAC E14.5 ${tissue}
height = 5
color = black
min_value = 0
max_value = ${max_value}
number_of_bins = 500
summary_method = max
[scale]
file_type = scalebar
x_center = 102754897
size = 1000
where = right
line_width = 3
[scale]
file_type = scalebar
x_center = 102838349
size = 5000
where = right
line_width = 3
"> ${output}.ini
pgt --tracks ${output}.ini --region chr15:102754066-102755729 -o pgt_outputs/${output}_EC1.svg --trackLabelFraction 0.3
pgt --tracks ${output}.ini --region chr15:102835414-102841286 -o pgt_outputs/${output}_EC2.svg --trackLabelFraction 0.3

